packages/core/src/compiler/Style.ts

```diff
@@ -3067,6 +3077,103 @@ const checkTranslation = (trans: Translation): StyleErrors => {
 
 //#endregion Checking translation
 
+export const checkCanvas = (tr: Translation): StyleErrors => {
+  let errs: StyleErrors = [];
+
+  /* TODO: type error reporting is wrong! */
+
+  if (!("canvas" in tr.trMap)) {
+    errs.push({
+      tag: "CanvasNonexistentError",
+    });
+
+    return errs;
+  }
+
+  if (!("width" in tr.trMap.canvas)) {
+    errs.push({
+      tag: "CanvasNonexistentDimsError",
+      attr: "width",
+      kind: "missing",
+    });
+  } else if (!("contents" in tr.trMap.canvas.width.contents)) {
+    errs.push({
+      tag: "CanvasNonexistentDimsError",
+      attr: "width",
+      kind: "GPI",
+    });
+  } else if (!("contents" in tr.trMap.canvas.width.contents.contents)) {
+    errs.push({
+      tag: "CanvasNonexistentDimsError",
+      attr: "width",
+      kind: "uninitialized",
+    });
+  } else if (
+    typeof tr.trMap.canvas.width.contents.contents.contents !== "number"
+  ) {
+    errs.push({
+      tag: "CanvasNonexistentDimsError",
+      attr: "width",
+      kind: "wrong type",
+      type: typeof tr.trMap.canvas.width.contents.contents.contents,
+    });
+  }
+
+  if (!("height" in tr.trMap.canvas)) {
+    errs.push({
+      tag: "CanvasNonexistentDimsError",
+      attr: "height",
+      kind: "missing",
+    });
+  } else if (!("contents" in tr.trMap.canvas.height.contents)) {
+    errs.push({
+      tag: "CanvasNonexistentDimsError",
+      attr: "height",
+      kind: "GPI",
+    });
+  } else if (!("contents" in tr.trMap.canvas.height.contents.contents)) {
+    errs.push({
+      tag: "CanvasNonexistentDimsError",
+      attr: "height",
+      kind: "uninitialized",
+    });
+  } else if (
+    typeof tr.trMap.canvas.height.contents.contents.contents !== "number"
+  ) {
+    const val = tr.trMap.canvas.height.contents.contents;
+    let type;
+    if (typeof val === "object" && "tag" in val) {
+      type = val.tag;
+    } else {
+      type = typeof val;
+    }
+
+    errs.push({
+      tag: "CanvasNonexistentDimsError",
+      attr: "height",
+      kind: "wrong type",
+      type,
+    });
+  }
+
+  return errs;
+};
+
+/* Precondition: checkCanvas returns without error */
+export const getCanvas = (tr: Translation): Canvas => {
```

Comment:
Yeah I normally try to avoid casts. The way to remove the casts here would be to add a `checkCanvas` call at the top and hope TypeScript can infer correctly based on that. However, `checkCanvas` is already run before any call to `getCanvas`.

Reply:
Ok so TS is unable to infer the correct types even when I add a `checkCanvas` call to `getCanvas`, so I think the casts are unavoidable.